segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,458 @@
+            syncWrite(handle, offset, data, length, timeout, retVal);
+        });
+
+        return retVal;
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        CompletableFuture<Void> retVal = new CompletableFuture<>();
+
+        executor.execute(() -> {
+            syncSeal(handle, timeout, retVal);
+        });
+
+        return retVal;
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+        CompletableFuture<Void> retVal = new CompletableFuture<>();
+
+        executor.execute(() -> {
+            syncConcat(targetHandle, offset, sourceSegment, timeout, retVal);
+        });
+
+        return retVal;
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+
+        executor.execute(() -> {
+            syncDelete(handle, timeout, future);
+        });
+
+        return future;
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    private void syncOpenRead(String streamSegmentName, CompletableFuture<SegmentHandle> retVal) {
+        Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            retVal.completeExceptionally(new StreamSegmentNotExistsException(streamSegmentName));
+            return;
+        }
+
+        FSSegmentHandle retHandle = FSSegmentHandle.getReadHandle(streamSegmentName);
+        retVal.complete(retHandle);
+    }
+
+
+    private void syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length, Duration
+            timeout, CompletableFuture<Integer> retVal) {
+
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (!Files.exists(path)) {
+            retVal.completeExceptionally(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+            return;
+        }
+
+        try {
+            if (Files.size(path) < offset) {
+                log.info("Read called on segment {} at offset {}. The offset is beyond the current size of the file.",
+                        handle.getSegmentName(), offset);
+                retVal.completeExceptionally(new ArrayIndexOutOfBoundsException());
+                return;
+            }
+        } catch (IOException e) {
+            retVal.completeExceptionally(e);
+            return;
+        }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
+            retVal.complete(bytesRead);
+        } catch (Exception e) {
+            if (e instanceof IndexOutOfBoundsException) {
```

Comment:
When would an IndexOutOfBounds occur? If it a possible code path we should probably check the bounds in advance.